README.md
# 相对于原作者，做了一些优化
1.边界崩溃问题

2.默认改为了人机对战

# QtWuziqi
Qt 写的五子棋小游戏，带AI和双人对战
# fork项目地址
https://github.com/tashaxing/QtWuziqi
# ScreenShot
![](https://github.com/tashaxing/QtWuziqi/raw/master/pic/wuziqi.gif)<br/>
# BlogAddress
http://blog.csdn.net/u012234115/article/details/53871009


debian/changelog
qtwuziqi (1.0.0) unstable; urgency=low

  * Initial release.

 -- Deepin Packages Builder <liuminghang0821@gmail.com>  Mon, 13 Feb 2017 16:32:34 +0800


GameModel.h
#ifndef GAMEMODEL_H
#define GAMEMODEL_H

// ---- 五子棋游戏模型类 ---- //
#include <QVector>

// 游戏类型，双人还是AI（目前固定让AI下黑子）
enum GameType {
    PERSON,
    BOT
};

// 游戏状态
enum GameStatus {
    PLAYING,
    WIN,
    DEAD
};

// 棋盘尺寸
const int kBoardSizeNum = 15;

class GameModel
{
public:
    GameModel();

public:
    QVector<QVector<int>> gameMapVec; // 存储当前游戏棋盘和棋子的情况,空白为0，白子1，黑子-1
    QVector<QVector<int>> scoreMapVec; // 存储各个点位的评分情况，作为AI下棋依据
    bool playerFlag; // 标示下棋方
    GameType gameType; // 游戏模式
    GameStatus gameStatus; // 游戏状态

    void startGame(GameType type); // 开始游戏
    void calculateScore(); // 计算评分
    void actionByPerson(int row, int col); // 人执行下棋
    void actionByAI(int &clickRow, int &clickCol); // 机器执行下棋
    void updateGameMap(int row, int col); // 每次落子后更新游戏棋盘
    bool isWin(int row, int col); // 判断游戏是否胜利
    bool isDeadGame(); // 判断是否和棋
};

#endif // GAMEMODEL_H


GameModel.cpp
#include <utility>
#include <stdlib.h>
#include <time.h>
#include "GameModel.h"

GameModel::GameModel()
{

}

void GameModel::startGame(GameType type)
{
    gameType = type;
    // 初始棋盘
    gameMapVec.clear();
    for (int i = 0; i < kBoardSizeNum; i++) {
        QVector<int> lineBoard;
        for (int j = 0; j < kBoardSizeNum; j++)
            lineBoard.push_back(0);
        gameMapVec.push_back(lineBoard);
    }

    // 如果是AI模式，需要初始化评分数组
    if (gameType == BOT) {
        scoreMapVec.clear();
        for (int i = 0; i < kBoardSizeNum; i++) {
            QVector<int> lineScores;
            for (int j = 0; j < kBoardSizeNum; j++)
                lineScores.push_back(0);
            scoreMapVec.push_back(lineScores);
        }
    }

    // 己方下为true,对方下位false
    playerFlag = true;


}

void GameModel::updateGameMap(int row, int col)
{
    if (playerFlag)
        gameMapVec[row][col] = 1;
    else
        gameMapVec[row][col] = -1;

    // 换手
    playerFlag = !playerFlag;
}

void GameModel::actionByPerson(int row, int col)
{
    updateGameMap(row, col);
}

void GameModel::actionByAI(int &clickRow, int &clickCol)
{
    // 计算评分
    calculateScore();

    // 从评分中找出最大分数的位置
    int maxScore = 0;
    QVector<std::pair<int, int>> maxPoints;

    for (int row = 1; row < kBoardSizeNum; row++)
        for (int col = 1; col < kBoardSizeNum; col++) {
            // 前提是这个坐标是空的
            if (gameMapVec[row][col] == 0) {
                if (scoreMapVec[row][col] > maxScore) {        // 找最大的数和坐标
                    maxPoints.clear();
                    maxScore = scoreMapVec[row][col];
                    maxPoints.push_back(std::make_pair(row, col));
                } else if (scoreMapVec[row][col] == maxScore)   // 如果有多个最大的数，都存起来
                    maxPoints.push_back(std::make_pair(row, col));
            }
        }

    // 随机落子，如果有多个点的话
    srand((unsigned)time(0));
    int index = rand() % maxPoints.size();

    std::pair<int, int> pointPair = maxPoints.at(index);
    clickRow = pointPair.first; // 记录落子点
    clickCol = pointPair.second;
    updateGameMap(clickRow, clickCol);
}

// 最关键的计算评分函数
void GameModel::calculateScore()
{
    // 统计玩家或者电脑连成的子
    int personNum = 0; // 玩家连成子的个数
    int botNum = 0; // AI连成子的个数
    int emptyNum = 0; // 各方向空白位的个数

    // 清空评分数组
    scoreMapVec.clear();
    for (int i = 0; i < kBoardSizeNum; i++) {
        QVector<int> lineScores;
        for (int j = 0; j < kBoardSizeNum; j++)
            lineScores.push_back(0);
        scoreMapVec.push_back(lineScores);
    }

    // 计分（此处是完全遍历，其实可以用bfs或者dfs加减枝降低复杂度，通过调整权重值，调整AI智能程度以及攻守风格）
    for (int row = 0; row < kBoardSizeNum; row++)
        for (int col = 0; col < kBoardSizeNum; col++) {
            // 空白点就算
            if (row > 0 && col > 0 &&
                    gameMapVec[row][col] == 0) {
                // 遍历周围八个方向
                for (int y = -1; y <= 1; y++)
                    for (int x = -1; x <= 1; x++) {
                        // 重置
                        personNum = 0;
                        botNum = 0;
                        emptyNum = 0;

                        // 原坐标不算
                        if (!(y == 0 && x == 0)) {
                            // 每个方向延伸4个子

                            // 对玩家白子评分（正反两个方向）
                            for (int i = 1; i <= 4; i++) {
                                if (row + i * y > 0 && row + i * y < kBoardSizeNum &&
                                        col + i * x > 0 && col + i * x < kBoardSizeNum &&
                                        gameMapVec[row + i * y][col + i * x] == 1) { // 玩家的子
                                    personNum++;
                                } else if (row + i * y > 0 && row + i * y < kBoardSizeNum &&
                                           col + i * x > 0 && col + i * x < kBoardSizeNum &&
                                           gameMapVec[row + i * y][col + i * x] == 0) { // 空白位
                                    emptyNum++;
                                    break;
                                } else          // 出边界
                                    break;
                            }

                            for (int i = 1; i <= 4; i++) {
                                if (row - i * y > 0 && row - i * y < kBoardSizeNum &&
                                        col - i * x > 0 && col - i * x < kBoardSizeNum &&
                                        gameMapVec[row - i * y][col - i * x] == 1) { // 玩家的子
                                    personNum++;
                                } else if (row - i * y > 0 && row - i * y < kBoardSizeNum &&
                                           col - i * x > 0 && col - i * x < kBoardSizeNum &&
                                           gameMapVec[row - i * y][col - i * x] == 0) { // 空白位
                                    emptyNum++;
                                    break;
                                } else          // 出边界
                                    break;
                            }

                            if (personNum == 1)                      // 杀二
                                scoreMapVec[row][col] += 10;
                            else if (personNum == 2) {               // 杀三
                                if (emptyNum == 1)
                                    scoreMapVec[row][col] += 30;
                                else if (emptyNum == 2)
                                    scoreMapVec[row][col] += 40;
                            } else if (personNum == 3) {             // 杀四
                                // 量变空位不一样，优先级不一样
                                if (emptyNum == 1)
                                    scoreMapVec[row][col] += 60;
                                else if (emptyNum == 2)
                                    scoreMapVec[row][col] += 110;
                            } else if (personNum == 4)               // 杀五
                                scoreMapVec[row][col] += 10100;

                            // 进行一次清空
                            emptyNum = 0;

                            // 对AI黑子评分
                            for (int i = 1; i <= 4; i++) {
                                if (row + i * y > 0 && row + i * y < kBoardSizeNum &&
                                        col + i * x > 0 && col + i * x < kBoardSizeNum &&
                                        gameMapVec[row + i * y][col + i * x] == 1) { // 玩家的子
                                    botNum++;
                                } else if (row + i * y > 0 && row + i * y < kBoardSizeNum &&
                                           col + i * x > 0 && col + i * x < kBoardSizeNum &&
                                           gameMapVec[row + i * y][col + i * x] == 0) { // 空白位
                                    emptyNum++;
                                    break;
                                } else          // 出边界
                                    break;
                            }

                            for (int i = 1; i <= 4; i++) {
                                if (row - i * y > 0 && row - i * y < kBoardSizeNum &&
                                        col - i * x > 0 && col - i * x < kBoardSizeNum &&
                                        gameMapVec[row - i * y][col - i * x] == -1) { // AI的子
                                    botNum++;
                                } else if (row - i * y > 0 && row - i * y < kBoardSizeNum &&
                                           col - i * x > 0 && col - i * x < kBoardSizeNum &&
                                           gameMapVec[row - i * y][col - i * x] == 0) { // 空白位
                                    emptyNum++;
                                    break;
                                } else          // 出边界
                                    break;
                            }

                            if (botNum == 0)                      // 普通下子
                                scoreMapVec[row][col] += 5;
                            else if (botNum == 1)                 // 活二
                                scoreMapVec[row][col] += 10;
                            else if (botNum == 2) {
                                if (emptyNum == 1)                // 死三
                                    scoreMapVec[row][col] += 25;
                                else if (emptyNum == 2)
                                    scoreMapVec[row][col] += 50;  // 活三
                            } else if (botNum == 3) {
                                if (emptyNum == 1)                // 死四
                                    scoreMapVec[row][col] += 55;
                                else if (emptyNum == 2)
                                    scoreMapVec[row][col] += 100; // 活四
                            } else if (botNum >= 4)
                                scoreMapVec[row][col] += 10000;   // 活五

                        }
                    }

            }
        }
}

bool GameModel::isWin(int row, int col)
{
    // 横竖斜四种大情况，每种情况都根据当前落子往后遍历5个棋子，有一种符合就算赢
    // 水平方向
    for (int i = 0; i < 5; i++) {
        // 往左5个，往右匹配4个子，20种情况
        if (col - i > 0 &&
                col - i + 4 < kBoardSizeNum &&
                gameMapVec[row][col - i] == gameMapVec[row][col - i + 1] &&
                gameMapVec[row][col - i] == gameMapVec[row][col - i + 2] &&
                gameMapVec[row][col - i] == gameMapVec[row][col - i + 3] &&
                gameMapVec[row][col - i] == gameMapVec[row][col - i + 4])
            return true;
    }

    // 竖直方向(上下延伸4个)
    for (int i = 0; i < 5; i++) {
        if (row - i > 0 &&
                row - i + 4 < kBoardSizeNum &&
                gameMapVec[row - i][col] == gameMapVec[row - i + 1][col] &&
                gameMapVec[row - i][col] == gameMapVec[row - i + 2][col] &&
                gameMapVec[row - i][col] == gameMapVec[row - i + 3][col] &&
                gameMapVec[row - i][col] == gameMapVec[row - i + 4][col])
            return true;
    }

    // 左斜方向
    for (int i = 0; i < 5; i++) {
        if (row + i < kBoardSizeNum &&
                row + i - 4 > 0 &&
                col - i > 0 &&
                col - i + 4 < kBoardSizeNum &&
                gameMapVec[row + i][col - i] == gameMapVec[row + i - 1][col - i + 1] &&
                gameMapVec[row + i][col - i] == gameMapVec[row + i - 2][col - i + 2] &&
                gameMapVec[row + i][col - i] == gameMapVec[row + i - 3][col - i + 3] &&
                gameMapVec[row + i][col - i] == gameMapVec[row + i - 4][col - i + 4])
            return true;
    }

    // 右斜方向
    for (int i = 0; i < 5; i++) {
        if (row - i > 0 &&
                row - i + 4 < kBoardSizeNum &&
                col - i > 0 &&
                col - i + 4 < kBoardSizeNum &&
                gameMapVec[row - i][col - i] == gameMapVec[row - i + 1][col - i + 1] &&
                gameMapVec[row - i][col - i] == gameMapVec[row - i + 2][col - i + 2] &&
                gameMapVec[row - i][col - i] == gameMapVec[row - i + 3][col - i + 3] &&
                gameMapVec[row - i][col - i] == gameMapVec[row - i + 4][col - i + 4])
            return true;
    }

    return false;
}

bool GameModel::isDeadGame()
{
    // 所有空格全部填满
    for (int i = 1; i < kBoardSizeNum; i++)
        for (int j = 1; j < kBoardSizeNum; j++) {
            if (!(gameMapVec[i][j] == 1 || gameMapVec[i][j] == -1))
                return false;
        }
    return true;
}



translations/QtWuziqi_zh_CN.ts
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="zh_CN">
<context>
    <name>MainWindow</name>
    <message>
        <location filename="../mainwindow.cpp" line="40"/>
        <source>Game</source>
        <translation>游戏</translation>
    </message>
    <message>
        <location filename="../mainwindow.cpp" line="41"/>
        <source>Person VS Person</source>
        <translation>人人对战</translation>
    </message>
    <message>
        <location filename="../mainwindow.cpp" line="45"/>
        <source>Person VS Computer</source>
        <translation>人机对战</translation>
    </message>
    <message>
        <location filename="../mainwindow.cpp" line="51"/>
        <source>QtWuziqi</source>
        <translation>Qt五子棋</translation>
    </message>
    <message>
        <location filename="../mainwindow.cpp" line="138"/>
        <source>white player win!</source>
        <translation>白棋胜利!</translation>
    </message>
    <message>
        <location filename="../mainwindow.cpp" line="140"/>
        <source>black player win!</source>
        <translation>黑棋胜利!</translation>
    </message>
    <message>
        <location filename="../mainwindow.cpp" line="142"/>
        <source>congratulations</source>
        <translation>恭喜</translation>
    </message>
</context>
</TS>


mainwindow.cpp
#include <QPainter>
#include <QTimer>
#include <QSound>
#include <QMouseEvent>
#include <QMessageBox>
#include <QMenuBar>
#include <QMenu>
#include <QAction>
#include <QDebug>
#include <math.h>
#include "mainwindow.h"

// -------全局遍历-------//
#define CHESS_ONE_SOUND ":/res/sound/chessone.wav"
#define WIN_SOUND ":/res/sound/win.wav"
#define LOSE_SOUND ":/res/sound/lose.wav"

const int kBoardMargin = 30; // 棋盘边缘空隙
const int kRadius = 15; // 棋子半径
const int kMarkSize = 6; // 落子标记边长
const int kBlockSize = 40; // 格子的大小
const int kPosDelta = 20; // 鼠标点击的模糊距离上限

const int kAIDelay = 700; // AI下棋的思考时间

// -------------------- //

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
{
    // 设置棋盘大小
    setFixedSize(kBoardMargin * 2 + kBlockSize * kBoardSizeNum, kBoardMargin * 2 + kBlockSize * kBoardSizeNum);
//    setStyleSheet("background-color:yellow;");

    // 开启鼠标hover功能，这两句一般要设置window的
    setMouseTracking(true);
//    centralWidget()->setMouseTracking(true);

    // 添加菜单
    QMenu *gameMenu = menuBar()->addMenu(tr("Game")); // menuBar默认是存在的，直接加菜单就可以了
    QAction *actionPVP = new QAction(tr("Person VS Person"), this);
    connect(actionPVP, SIGNAL(triggered()), this, SLOT(initPVPGame()));
    gameMenu->addAction(actionPVP);

    QAction *actionPVE = new QAction(tr("Person VS Computer"), this);
    connect(actionPVE, SIGNAL(triggered()), this, SLOT(initPVEGame()));
    gameMenu->addAction(actionPVE);

    // 开始游戏
    initGame();
    setWindowTitle(tr("QtWuziqi"));
}

MainWindow::~MainWindow()
{
    if (game) {
        delete game;
        game = nullptr;
    }
}

void MainWindow::initGame()
{
    // 初始化游戏模型
    game = new GameModel;
    initPVEGame();
}

void MainWindow::initPVPGame()
{
    game_type = PERSON;
    game->gameStatus = PLAYING;
    game->startGame(game_type);
    update();
}

void MainWindow::initPVEGame()
{
    game_type = BOT;
    game->gameStatus = PLAYING;
    game->startGame(game_type);
    update();
}

void MainWindow::paintEvent(QPaintEvent *event)
{
    QPainter painter(this);
    // 绘制棋盘
    painter.setRenderHint(QPainter::Antialiasing, true); // 抗锯齿
//    QPen pen; // 调整线条宽度
//    pen.setWidth(2);
//    painter.setPen(pen);
    for (int i = 0; i < kBoardSizeNum + 1; i++) {
        painter.drawLine(kBoardMargin + kBlockSize * i, kBoardMargin, kBoardMargin + kBlockSize * i, size().height() - kBoardMargin);
        painter.drawLine(kBoardMargin, kBoardMargin + kBlockSize * i, size().width() - kBoardMargin, kBoardMargin + kBlockSize * i);
    }

    QBrush brush;
    brush.setStyle(Qt::SolidPattern);
    // 绘制落子标记(防止鼠标出框越界)
    if (clickPosRow > 0 && clickPosRow < kBoardSizeNum &&
            clickPosCol > 0 && clickPosCol < kBoardSizeNum &&
            game->gameMapVec[clickPosRow][clickPosCol] == 0) {
        if (game->playerFlag)
            brush.setColor(Qt::white);
        else
            brush.setColor(Qt::black);
        painter.setBrush(brush);
        painter.drawRect(kBoardMargin + kBlockSize * clickPosCol - kMarkSize / 2, kBoardMargin + kBlockSize * clickPosRow - kMarkSize / 2, kMarkSize, kMarkSize);
    }

    // 绘制棋子
    for (int i = 0; i < kBoardSizeNum; i++)
        for (int j = 0; j < kBoardSizeNum; j++) {
            if (game->gameMapVec[i][j] == 1) {
                brush.setColor(Qt::white);
                painter.setBrush(brush);
                painter.drawEllipse(kBoardMargin + kBlockSize * j - kRadius, kBoardMargin + kBlockSize * i - kRadius, kRadius * 2, kRadius * 2);
            } else if (game->gameMapVec[i][j] == -1) {
                brush.setColor(Qt::black);
                painter.setBrush(brush);
                painter.drawEllipse(kBoardMargin + kBlockSize * j - kRadius, kBoardMargin + kBlockSize * i - kRadius, kRadius * 2, kRadius * 2);
            }
        }

    // 判断输赢
    if (clickPosRow > 0 && clickPosRow < kBoardSizeNum &&
            clickPosCol > 0 && clickPosCol < kBoardSizeNum &&
            (game->gameMapVec[clickPosRow][clickPosCol] == 1 ||
             game->gameMapVec[clickPosRow][clickPosCol] == -1)) {
        if (game->isWin(clickPosRow, clickPosCol) && game->gameStatus == PLAYING) {
            qDebug() << "win";
            game->gameStatus = WIN;
            QSound::play(WIN_SOUND);
            QMessageBox::StandardButton btnValue;
            QString str;
            if (game->gameMapVec[clickPosRow][clickPosCol] == 1) {
                str = tr("white player win!");
            } else if (game->gameMapVec[clickPosRow][clickPosCol] == -1) {
                str = tr("black player win!");
            }
            btnValue = QMessageBox::information(this, tr("congratulations"), str);
            // 重置游戏状态，否则容易死循环
            if (btnValue == QMessageBox::Ok) {
                game->startGame(game_type);
                game->gameStatus = PLAYING;
            }
        }
    }


    // 判断死局
    if (game->isDeadGame()) {
        QSound::play(LOSE_SOUND);
        QMessageBox::StandardButton btnValue = QMessageBox::information(this, "oops", "dead game!");
        if (btnValue == QMessageBox::Ok) {
            game->startGame(game_type);
            game->gameStatus = PLAYING;
        }

    }
}

void MainWindow::mouseMoveEvent(QMouseEvent *event)
{
    // 通过鼠标的hover确定落子的标记
    int x = event->x();
    int y = event->y();

    // 棋盘边缘不能落子
    if (x >= kBoardMargin + kBlockSize / 2 &&
            x < size().width() - kBoardMargin &&
            y >= kBoardMargin + kBlockSize / 2 &&
            y < size().height() - kBoardMargin) {
        // 获取最近的左上角的点
        int col = x / kBlockSize;
        int row = y / kBlockSize;

        int leftTopPosX = kBoardMargin + kBlockSize * col;
        int leftTopPosY = kBoardMargin + kBlockSize * row;

        // 根据距离算出合适的点击位置,一共四个点，根据半径距离选最近的
        clickPosRow = -1; // 初始化最终的值
        clickPosCol = -1;
        int len = 0; // 计算完后取整就可以了

        // 确定一个误差在范围内的点，且只可能确定一个出来
        len = sqrt((x - leftTopPosX) * (x - leftTopPosX) + (y - leftTopPosY) * (y - leftTopPosY));
        if (len < kPosDelta) {
            clickPosRow = row;
            clickPosCol = col;
        }
        len = sqrt((x - leftTopPosX - kBlockSize) * (x - leftTopPosX - kBlockSize) + (y - leftTopPosY) * (y - leftTopPosY));
        if (len < kPosDelta) {
            clickPosRow = row;
            clickPosCol = col + 1;
        }
        len = sqrt((x - leftTopPosX) * (x - leftTopPosX) + (y - leftTopPosY - kBlockSize) * (y - leftTopPosY - kBlockSize));
        if (len < kPosDelta) {
            clickPosRow = row + 1;
            clickPosCol = col;
        }
        len = sqrt((x - leftTopPosX - kBlockSize) * (x - leftTopPosX - kBlockSize) + (y - leftTopPosY - kBlockSize) * (y - leftTopPosY - kBlockSize));
        if (len < kPosDelta) {
            clickPosRow = row + 1;
            clickPosCol = col + 1;
        }
    }

    // 存了坐标后也要重绘
    update();
}

void MainWindow::mouseReleaseEvent(QMouseEvent *event)
{
    // 人下棋，并且不能抢机器的棋
    if (!(game_type == BOT && !game->playerFlag)) {
        chessOneByPerson();
        // 如果是人机模式，需要调用AI下棋
        if (game->gameType == BOT && !game->playerFlag) {
            // 用定时器做一个延迟
            QTimer::singleShot(kAIDelay, this, SLOT(chessOneByAI()));
        }
    }

}

void MainWindow::chessOneByPerson()
{
    // 根据当前存储的坐标下子
    // 只有有效点击才下子，并且该处没有子
    if (clickPosRow != -1 && clickPosCol != -1
            && clickPosRow < game->gameMapVec.size()
            && clickPosCol < game->gameMapVec[clickPosRow].size()
            && game->gameMapVec[clickPosRow][clickPosCol] == 0) {
        game->actionByPerson(clickPosRow, clickPosCol);
        QSound::play(CHESS_ONE_SOUND);

        // 重绘
        update();
    }
}

void MainWindow::chessOneByAI()
{
    game->actionByAI(clickPosRow, clickPosCol);
    QSound::play(CHESS_ONE_SOUND);
    update();
}

